keps/sig-node/127-usernamespaces-support/README.md

```diff
@@ -0,0 +1,788 @@
+
+This proposal implementation aims to be divided in the following phases:
+
+### Phase 1
+
+This first phase includes:
+ - Extend the PodSpec with the `userNamespaceMode` field.
+ - Extend the CRI with user and ID mappings fields.
+ - Implement support for `Host` and `Cluster` user namespace modes.
+
+The goal of this phase is to implement some initial user namespace support
+providing pod-to-host isolation and supporting volumes. The implementation of
+the `Pod` mode is out of scope in this phase because it requires a non
+negligible amount of work and we could risk losing the focus failing to deliver
+this feature.
+
+### Phase 2
+
+This phase aims to implement the `Pod` mode. After this phase is completed the
+full advantanges of user namespaces could be used in some cases (stateless
+workloads).
+
+### Phase 2+
+
+The default value of `userNamespaceMode` should be set to `Pod` so pods that
+don't set this field can also get the security benefits of user namespaces. It's
+not clear yet what should be the process to make this happen as this is a
+potentially non backwards compatible change. It's specially relevant for
+workloads not compatible with user namespaces.
+
+A [host defaulting mechanishm](#host-defaulting-mechanishm) could help to make
+this transiction smoother, but this proposal doesn't go into details as it
+mainly focuses in phase 1.
+
+TODO(Mauricio):
+- Should we describe that once the default is Pod we should implement a control for it on PSP / OPA?
+- Should we mention that it's possible that users in future would be able to set the mappings?
+
+## Design Details
+
+This section only focuses on phase 1 as specified above.
+
+### Summary of the Proposed Changes
+
+- Extend the CRI to have a user namespace mode and the user and group ID mappings.
+- Add a `userNamespaceMode` field to the pod spec.
+- Add the cluster-wide ID mappings to the kubelet configuration file.
+- Add a `UserNamespacesSupport` feature flag to enable / disable the user.
+  namespaces support.
+- Update owner of ephemeral volumes populated by the kubelet.
+
+### PodSpec Changes
+
+`v1.PodSpec` is extended with a new `UserNamesapceMode` field:
+
+```
+const (
+	UserNamespaceModeHost    PodUserNamespaceMode = "Host"
+	UserNamespaceModeCluster PodUserNamespaceMode = "Cluster"
+	UserNamespaceModePod     PodUserNamespaceMode = "Pod"
+)
+
+type PodSpec struct {
+...
+  // UserNamespaceMode controls how user namespaces are used for this Pod.
+  // Three modes are supported:
+  // "Host": The pod shares the host user namespace. (default value).
+  // "Cluster": The pod uses a cluster-wide configured ID mappings.
+  // "Pod": The pod gets a non-overlapping ID mappings range.
+  // +k8s:conversion-gen=false
+  // +optional
+  UserNamespaceMode PodUserNamespaceMode `json:"userNamespaceMode,omitempty" protobuf:"bytes,36,opt  name=userNamespaceMode"`
+...
+```
+
+### CRI API Changes
+
+The CRI has to be extended to allow kubelet to specify the user namespace mode
```

Comment:
Just as an example of my previous comment. Something like this I think is clearer in general (I know in spanish it isn't, but in English I think it is :)):

```suggestion
The CRI is extended to (optionally) specify the user namespace mode
```

Reply:
Right. I'll do a pass improving the verb tense. It's a mix between present and future conditional now.